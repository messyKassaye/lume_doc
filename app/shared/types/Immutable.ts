import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
