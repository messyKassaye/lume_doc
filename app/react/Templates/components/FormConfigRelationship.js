import { Field } from 'react-redux-form';
import { connect } from 'react-redux';
import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { createSelector } from 'reselect';

import { Select } from 'app/ReactReduxForms';
import { Translate } from 'app/I18N';
import { Icon } from 'app/UI';
import PropertyConfigOptions from './PropertyConfigOptions';
import PropertyConfigOption from './PropertyConfigOption';
import Tip from '../../Layout/Tip';

export class FormConfigRelationship extends Component {
  static contentValidation() {
    return { required: val => val && val.trim() !== '' };
  }

  render() {
    const {
      index,
      type,
      inheritPropertyError,
      labelError,
      relationTypeError,
      templates,
      relationTypes,
      templateId,
      showInheritOption,
      showInheritSelect,
      templateProperties,
      inheritSelectPropertyType,
    } = this.props;

    const options = templates.toJS().filter(template => template._id !== templateId);

    const labelClass = labelError ? 'form-group has-error' : 'form-group';

    return (
      <div>
        <div className={labelClass}>
          <label htmlFor="label">
            <Translate>Label</Translate>
          </label>
          <Field model={`template.data.properties[${index}].label`}>
            <input id="label" className="form-control" />
          </Field>
        </div>
        <div className={relationTypeError ? 'form-group has-error' : 'form-group'}>
          <label>
            <Translate>Relationship</Translate>
            <span className="required">*</span>
          </label>
          <Select
            model={`template.data.properties[${index}].relationType`}
            options={relationTypes.toJS()}
            optionsLabel="name"
            validators={FormConfigRelationship.contentValidation()}
            optionsValue="_id"
          />
        </div>
        <div className="form-group">
          <label>
            <Translate>Entities</Translate>
          </label>
          <Select
            model={`template.data.properties[${index}].content`}
            options={options}
            optionsLabel="name"
            placeholder="Any entity or document"
            optionsValue="_id"
          />
        </div>
        {showInheritOption && (
          <PropertyConfigOption
            label="Inherit property"
            model={`template.data.properties[${index}].inherit`}
          >
            <Tip>
              This property will be inherited from the related entities and shown as metadata of
              this type of entities.
            </Tip>
          </PropertyConfigOption>
        )}
        {showInheritSelect && (
          <div className={inheritPropertyError ? 'form-group has-error' : 'form-group'}>
            <Select
              model={`template.data.properties[${index}].inheritProperty`}
              options={templateProperties}
              optionsLabel="label"
              optionsValue="_id"
            />
          </div>
        )}
        {showInheritSelect && inheritSelectPropertyType === 'geolocation' && (
          <div className="geolocation-grouping-alert">
            <Icon icon="info-circle" />
            <p>
              <Translate>Adjacent geolocation properties will render on the same map</Translate>.
            </p>
          </div>
        )}
        <PropertyConfigOptions index={index} type={type} />
      </div>
    );
  }
}

FormConfigRelationship.defaultProps = {
  labelError: false,
  relationTypeError: false,
  inheritPropertyError: false,
  showInheritOption: false,
  showInheritSelect: false,
  templateId: null,
  inheritSelectPropertyType: null,
};

FormConfigRelationship.propTypes = {
  templates: PropTypes.instanceOf(Immutable.List).isRequired,
  relationTypes: PropTypes.instanceOf(Immutable.List).isRequired,
  templateProperties: PropTypes.array.isRequired,
  index: PropTypes.number.isRequired,
  type: PropTypes.string.isRequired,
  labelError: PropTypes.bool,
  relationTypeError: PropTypes.bool,
  inheritPropertyError: PropTypes.bool,
  showInheritOption: PropTypes.bool,
  showInheritSelect: PropTypes.bool,
  templateId: PropTypes.string,
  inheritSelectPropertyType: PropTypes.string,
};

const getTemplateProperties = createSelector(
  state => state.templates,
  (state, props) =>
    state.template.formState.properties[props.index].content
      ? state.template.formState.properties[props.index].content.value
      : null,
  (templates, content) => {
    const targetTemplate = templates.find(t => t.get('_id') === content);
    return targetTemplate ? targetTemplate.get('properties').toJS() : [];
  }
);

const getInheritSelectPropertyType = createSelector(
  getTemplateProperties,
  (state, props) => state.template.data.properties[props.index].inheritProperty,
  (templateProperties, inheritedPropertyId) => {
    const inheritedProperty = templateProperties.find(p => p._id === inheritedPropertyId);
    return inheritedProperty && inheritedProperty.type;
  }
);

export function mapStateToProps(state, props) {
  const { template, templates, relationTypes } = state;

  const templateProperties = getTemplateProperties(state, props);

  return {
    labelError:
      template.formState.$form.errors[`properties.${props.index}.label.required`] ||
      template.formState.$form.errors[`properties.${props.index}.label.duplicated`],

    relationTypeError:
      template.formState.$form.errors[`properties.${props.index}.relationType.required`] &&
      template.formState.$form.submitFailed,

    inheritPropertyError:
      template.formState.$form.errors[`properties.${props.index}.inheritProperty.required`] &&
      template.formState.$form.submitFailed,

    showInheritOption: Boolean(
      template.formState.properties[props.index].content && templateProperties.length
    ),

    showInheritSelect: Boolean(
      template.formState.properties[props.index].inherit &&
        template.formState.properties[props.index].inherit.value &&
        templateProperties.length
    ),
    inheritSelectPropertyType: getInheritSelectPropertyType(state, props),
    templateProperties,
    templateId: template.data._id,
    templates,
    relationTypes,
  };
}

export default connect(mapStateToProps)(FormConfigRelationship);
