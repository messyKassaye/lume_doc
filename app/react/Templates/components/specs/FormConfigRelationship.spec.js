import Immutable from 'immutable';

import FormConfigRelationship from '../FormConfigRelationship';
import { renderConnected } from '../../specs/utils/renderConnected.tsx';

describe('FormConfigRelationship', () => {
  let templates;
  let relationTypes;
  let props;
  let storeData;

  const render = () => renderConnected(FormConfigRelationship, props, storeData);

  beforeEach(() => {
    templates = [{ _id: 3, name: 'Judge', type: 'template', properties: [] }];
    relationTypes = [
      { _id: 1, name: 'relationType1' },
      { _id: 2, name: 'relationType2' },
    ];

    storeData = {
      templates: Immutable.fromJS(templates),
      relationTypes: Immutable.fromJS(relationTypes),
      template: {
        data: { properties: [{ filter: false }] },
        formState: {
          'properties.0.label': { valid: true, dirty: false, errors: {} },
          properties: [{ content: { value: 3 } }],
          $form: {
            errors: {},
          },
        },
      },
    };

    props = {
      index: 0,
      type: 'relationship',
    };
  });

  it('should render fields with the correct datas', () => {
    const component = render();
    expect(component).toMatchSnapshot();
  });

  describe('when the fields are invalid and dirty or the form is submited', () => {
    it('should render the label with errors', () => {
      storeData.template.formState.$form.errors['properties.0.label.required'] = true;

      const component = render();
      expect(component).toMatchSnapshot();
    });

    it('should render the label with errors when the form is submited', () => {
      storeData.template.formState.$form.errors['properties.0.label.required'] = true;
      storeData.template.formState.submitFailed = true;

      const component = render();
      expect(component).toMatchSnapshot();
    });
  });

  describe('when use as filter is selected', () => {
    it('should show the default filter option', () => {
      storeData.template.data.properties[0].filter = true;

      const component = render();
      expect(component).toMatchSnapshot();
    });
  });
});
