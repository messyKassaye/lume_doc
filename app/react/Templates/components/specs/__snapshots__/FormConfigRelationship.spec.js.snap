// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormConfigRelationship should render fields with the correct datas 1`] = `
<div>
  <div
    className="form-group"
  >
    <label
      htmlFor="label"
    >
      <Connect(Translate)>
        Label
      </Connect(Translate)>
    </label>
    <Modeled(Connect(Field))
      model="template.data.properties[0].label"
    >
      <input
        className="form-control"
        id="label"
      />
    </Modeled(Connect(Field))>
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Relationship
      </Connect(Translate)>
      <span
        className="required"
      >
        *
      </span>
    </label>
    <Select
      model="template.data.properties[0].relationType"
      options={
        Array [
          Object {
            "_id": 1,
            "name": "relationType1",
          },
          Object {
            "_id": 2,
            "name": "relationType2",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      validators={
        Object {
          "required": [Function],
        }
      }
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Entities
      </Connect(Translate)>
    </label>
    <Select
      model="template.data.properties[0].content"
      options={
        Array [
          Object {
            "_id": 3,
            "name": "Judge",
            "properties": Array [],
            "type": "template",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      placeholder="Any entity or document"
    />
  </div>
  <Connect(PropertyConfigOptions)
    index={0}
    type="relationship"
  />
</div>
`;

exports[`FormConfigRelationship when the fields are invalid and dirty or the form is submited should render the label with errors 1`] = `
<div>
  <div
    className="form-group has-error"
  >
    <label
      htmlFor="label"
    >
      <Connect(Translate)>
        Label
      </Connect(Translate)>
    </label>
    <Modeled(Connect(Field))
      model="template.data.properties[0].label"
    >
      <input
        className="form-control"
        id="label"
      />
    </Modeled(Connect(Field))>
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Relationship
      </Connect(Translate)>
      <span
        className="required"
      >
        *
      </span>
    </label>
    <Select
      model="template.data.properties[0].relationType"
      options={
        Array [
          Object {
            "_id": 1,
            "name": "relationType1",
          },
          Object {
            "_id": 2,
            "name": "relationType2",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      validators={
        Object {
          "required": [Function],
        }
      }
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Entities
      </Connect(Translate)>
    </label>
    <Select
      model="template.data.properties[0].content"
      options={
        Array [
          Object {
            "_id": 3,
            "name": "Judge",
            "properties": Array [],
            "type": "template",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      placeholder="Any entity or document"
    />
  </div>
  <Connect(PropertyConfigOptions)
    index={0}
    type="relationship"
  />
</div>
`;

exports[`FormConfigRelationship when the fields are invalid and dirty or the form is submited should render the label with errors when the form is submited 1`] = `
<div>
  <div
    className="form-group has-error"
  >
    <label
      htmlFor="label"
    >
      <Connect(Translate)>
        Label
      </Connect(Translate)>
    </label>
    <Modeled(Connect(Field))
      model="template.data.properties[0].label"
    >
      <input
        className="form-control"
        id="label"
      />
    </Modeled(Connect(Field))>
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Relationship
      </Connect(Translate)>
      <span
        className="required"
      >
        *
      </span>
    </label>
    <Select
      model="template.data.properties[0].relationType"
      options={
        Array [
          Object {
            "_id": 1,
            "name": "relationType1",
          },
          Object {
            "_id": 2,
            "name": "relationType2",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      validators={
        Object {
          "required": [Function],
        }
      }
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Entities
      </Connect(Translate)>
    </label>
    <Select
      model="template.data.properties[0].content"
      options={
        Array [
          Object {
            "_id": 3,
            "name": "Judge",
            "properties": Array [],
            "type": "template",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      placeholder="Any entity or document"
    />
  </div>
  <Connect(PropertyConfigOptions)
    index={0}
    type="relationship"
  />
</div>
`;

exports[`FormConfigRelationship when use as filter is selected should show the default filter option 1`] = `
<div>
  <div
    className="form-group"
  >
    <label
      htmlFor="label"
    >
      <Connect(Translate)>
        Label
      </Connect(Translate)>
    </label>
    <Modeled(Connect(Field))
      model="template.data.properties[0].label"
    >
      <input
        className="form-control"
        id="label"
      />
    </Modeled(Connect(Field))>
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Relationship
      </Connect(Translate)>
      <span
        className="required"
      >
        *
      </span>
    </label>
    <Select
      model="template.data.properties[0].relationType"
      options={
        Array [
          Object {
            "_id": 1,
            "name": "relationType1",
          },
          Object {
            "_id": 2,
            "name": "relationType2",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      validators={
        Object {
          "required": [Function],
        }
      }
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      <Connect(Translate)>
        Entities
      </Connect(Translate)>
    </label>
    <Select
      model="template.data.properties[0].content"
      options={
        Array [
          Object {
            "_id": 3,
            "name": "Judge",
            "properties": Array [],
            "type": "template",
          },
        ]
      }
      optionsLabel="name"
      optionsValue="_id"
      placeholder="Any entity or document"
    />
  </div>
  <Connect(PropertyConfigOptions)
    index={0}
    type="relationship"
  />
</div>
`;
